### PR TITLE
Support dense density matrices in expectation()

### PR DESCRIPTION
**Pull request by mpharrigan**

### CODE REVIEW

File: src/openfermion/utils/_sparse_tools.py [hunk 1]
```diff
@@ -744,32 +744,36 @@ def expectation(operator, state):
...
         if isinstance(operator, scipy.sparse.linalg.LinearOperator):
             raise ValueError('Taking the expectation of a LinearOperator with '
                              'a density matrix is not supported.')
         product = state * operator
-        expectation = numpy.sum(product.diagonal())
+        return numpy.sum(product.diagonal())
 
-    elif isinstance(state, numpy.ndarray):
+    if isinstance(state, numpy.ndarray):
         # Handle state vector.
```
Comment (kevinsung) [inline src/openfermion/utils/_sparse_tools.py:756]: Should remove this comment now that it includes a density matrix case.

Comment (mpharrigan) [reply src/openfermion/utils/_sparse_tools.py]: done

**Comment by mpharrigan** [comment]
ptal

File: src/openfermion/utils/_sparse_tools.py [hunk 2]
```diff
@@ -744,32 +744,34 @@ def expectation(operator, state):
...
+            return numpy.dot(numpy.conjugate(state), operator * state)
+        if len(state.shape) == 2:
+            if state.shape[1] == 1:
+                # Column vector
+                return numpy.dot(numpy.conjugate(state.T),
+                                 operator * state).item()
+            if state.shape[0] == state.shape[1]:
+                # Density matrix
+                product = state * operator
+                return numpy.sum(product.diagonal()).item()
```
Comment (kevinsung) [inline src/openfermion/utils/_sparse_tools.py:767]: You can just use `numpy.trace`

Comment (mpharrigan) [reply src/openfermion/utils/_sparse_tools.py:767]: aha! you would think that. it does not work on sparse matrices

Comment (kevinsung) [reply src/openfermion/utils/_sparse_tools.py:767]: Oh

Comment (kevinsung) [reply src/openfermion/utils/_sparse_tools.py:767]: But multiplying a sparse matrix with a numpy array gives a numpy array. So here, `product` should be a numpy array, right?

Comment (ncrubin) [reply src/openfermion/utils/_sparse_tools.py:767]: looks like this is correct.  state is a sparse matrix and operator is a sparse matrix.

Comment (kevinsung) [reply src/openfermion/utils/_sparse_tools.py:767]: `state` is a Numpy array (see the if statement in line 755).

**Comment by ncrubin** [comment]
@kevinsung what's the status of this?

**Comment by mpharrigan** [comment]
The status is I forgot about it. This is sortof low on my priority list right now

**Comment by obriente** [comment]
Closing for now (feel free to reopen if you can find the time).

**Pull request closed without merging**